### PR TITLE
golden gen for decoder

### PR DESCRIPTION
**Pull request by Buck008**
I've used Numpy and referenced code from NanoGPT to write the decoder forward dataflow based on my understanding. Currently, all values are floating-point numbers. (All weights are randomly generated, and I haven't implemented the dropout function because I believe it won't be executed in hardware.) I would greatly appreciate it if someone could help me check whether my understanding is correct. (Please excuse my limited Python skills.) If this is right, I am going to write the integer version,which will be the golden brick generator.

### CODE REVIEW

File: tf_np_golden_gen/forward.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+y = y_projected + attn_input_x
+MLP_input_x = y
+y = layernorm(y,gamma, beta)
+
+#MLP
+
+W_fc = np.random.randn(C, 4 * C)
+b_fc = np.random.randn(4 * C)
+W_proj = np.random.randn(4 * C, C)
+b_proj = np.random.randn(C)
```
Comment (gkielian) [inline tf_np_golden_gen/forward.py:88]: Can remove b_fc and b_proj too

File: tf_np_golden_gen/forward.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
+import numpy as np
+
+
+def layernorm(x, gamma, beta, epsilon=1e-5):
+    mean = np.mean(x, axis=-1, keepdims=True)
+    variance = np.var(x, axis=-1, keepdims=True)
+    x_normalized = (x - mean) / np.sqrt(variance + epsilon)
+    y = gamma * x_normalized + beta
```
Comment (gkielian) [inline tf_np_golden_gen/forward.py:8]: Let's try to emulate RMSNorm -- which essentially just assumes the mean is zero and that beta is a zero tensor.

One way to do this is to add a named option to the layernorm signature term for setting mean to zero, example:

```
def layernorm(x, gamma, beta, espsilon=1e-5, mean_zero=True):
```

**Comment by Buck008** [comment]
Sure, I will do all these above.  I am not familar with embedding for nano gpt now, I may need time to see that. And the embeding layer now is not supported in hardware.

**Comment by Buck008** [comment]
Sorry I don't understand how the embedding layer do in the network, so I just create the two matrixes. And for the output layer, do you mean a layernorm or RMSnorm? For hardware, we are just going to support attention and mlp.

**Comment by gkielian** [comment]
> Sure, I will do all these above. I am not familar with embedding for nano gpt now, I may need time to see that. And the embeding layer now is not supported in hardware.

Good to know, I think we're closing in on a solution for the embedding layer for hardware too

**Review by gkielian — APPROVED** [review]
Looks great for a first commit : )

Merging

**Pull request merged**